## 错误边界

#### 1. 背景

React16对待组件渲染错误的策略是，任何未被捕获的错误将会导致整个React组件树整个被卸载。

因此React提供了一种机制让开发者捕获子组件树中的错误。这就是错误边界。

#### 2. 错误边界相关的API

- 如果一个class组件定义了```static getDerivedStateFromError```或者```componentDidCatch```中的任何一个，它就变成一个错误边界
- 当抛出错误后，使用```static DerivedStateFromError```渲染备用UI，使用```componentDidCatch```打印信息。

#### 3. 错误边界捕获的信息

错误边界捕获的是其子组件树的错误，它不能捕获

- 事件处理
- 异步代码
- 服务端渲染
- 它自身抛出的错误

【思考】由于异常捕获是try catch子组件生命周期钩子中的错误，因此上述情况无法捕获。

#### 4. 错误边界应该放在哪里

大多数情况下, 你只需要声明一次错误边界组件, 并在整个应用中使用它。

- 放在全局，以告诉开发者具体的子组件的错误
- 可以将单独的部件（如第三方React组件）包在错误边界中，以保证不会影响到其他组件

#### 5. 组件栈追踪

React16会把渲染期间发生的所有错误打印到控制台。

打印的信息除了错误信息和JavaScript调用堆栈外，React还支持组件栈追踪。即查看发生错误的组件名和行号。

组件栈错误追踪的功能需要相关支持，React需要知道： 1. 该组件的名称和 2. 源代码行号。这是通过一些第三方工具实现的

1. 目前主流浏览器都支持function.name，这样React可以通过类的name属性知道组件的名称。对于不支持function.name的浏览器，可以通过```function.name-polyfill```或者组件上挂载```displayName```属性来实现。

2. 使用babel插件[babel-plugin-transform-react-jsx-source
](https://www.npmjs.com/package/babel-plugin-transform-react-jsx-source)给组件添加```__source```属性

#### 6. 示例代码

```
// 定义错误边界
class ErrorBoundary extends React.Component {
  constructor(props) {
    super(props);
    this.state = { hasError: false };
  }

  static getDerivedStateFromError(error) {
    // 更新 state 使下一次渲染能够显示降级后的 UI
    return { hasError: true };
  }

  componentDidCatch(error, info) {
    // 你同样可以将错误日志上报给服务器
    logErrorToMyService(error, info);
  }

  render() {
    if (this.state.hasError) {
      // 你可以自定义降级后的 UI 并渲染
      return <h1>Something went wrong.</h1>;
    }

    return this.props.children; 
  }
}

// 使用错误边界
<ErrorBoundary>
  <MyWidget />
</ErrorBoundary>
```